design/implicit-boot-mode.md

```diff
@@ -0,0 +1,197 @@
+are `bios` and `uefi`. When the `boot_mode` property is
+explicitly set, Ironic will try to ensure that the hardware is
+configured accordingly and it may also choose different paths in the
+provisioning process. In particular, there is a branch in the ISO
+building logic to include different content based on the boot mode.
+
+If the `boot_mode` is not specified, Ironic tries to determine the
+right value by looking at the hardware settings. Some BMC types do not
+support this capability, especially when using the IPMI protocol. When
+Ironic cannot determine the right value for itself, and has not been
+told explicitly what value to use, it falls back to a global default
+specified by a configuration option. Metal3 is not typically deployed
+in a way that allows that option to be changed easily.
+
+We want to ensure that metal3's behavior is always deterministic and
+predictable. We also want to keep the API as minimal as possible, and
+to avoid introducing API fields in a way that makes it likely that a
+user would either need to tell us something we can figure out or that
+they could tell us something incorrect.
+
+Given those goals, it is useful to examine the different situations in
+which different boot modes are appropriate, and how to select the
+right mode at the right time.
+
+The first approach would be to try to always require the same boot
+mode for all hosts. Most modern x86 hosts support UEFI. It is
+therefore useful as a default. However, many contributors have
+somewhat older hardware, which works better with the BIOS boot
+mode. While this hardware is not necessarily our target, it isn't
+practical to abandon support entirely by always requiring
+UEFI. Similarly, there are reasons we cannot always require a host to
+use the BIOS boot mode. UEFI supports desirable features, such as
+secure booting, non-x86 hardware, and PXE on IPv6 networks. For these
+reasons, a single boot mode will not work.
+
+The desired features do give us some hints about how we can determine
+the boot mode implicitly, however. For example, we have no plans to
+support virtual media with IPMI. We can also tell whether the
+provisioning network is using IPv4 or IPv6. Using this information, we
+could build a set of rules to determine the boot mode. However, we
+would have to document those rules, and they may not seem obvious to
+the end-user.
+
+At least some of the features are tied to the protocol used to
+communicate with the BMC. We could therefore link the boot mode to the
+BMC driver, and have a deterministic method for choosing the boot mode
+that is also easy to document.
+
+The proposal, therefore, is to always use BIOS mode when using IPMI
+and to prefer UEFI when using Redfish. We can implement this by adding
+a new method to the [AccessDetails
+interface](https://github.com/metal3-io/baremetal-operator/blob/master/pkg/bmc/access.go#L27)
+to return the properties for a host, including the `boot_mode`. This
```

Comment:
When we say return host ```properties```, are we refering to the Ironic Node ```capabilities```? (I understand we would want to name them differently).

Reply:
I think we want it to return a map[string]interface{} that can be passed as the Properties field of a CreateOpts when registering the node with ironic. See https://github.com/gophercloud/gophercloud/blob/master/openstack/baremetal/v1/nodes/requests.go#L220

Reply:
I see. Okay.